libmscore/scale.cpp

```diff
@@ -0,0 +1,265 @@
+
+      fillMissing12();
+      }
+
+//---------------------------------------------------------
+//   fillMissing12
+//---------------------------------------------------------
+void Scale::fillMissing12() {
+      computedNotes[TPC_D_B] = computedNotes[TPC_C_S];
+      computedNotes[TPC_E_B] = computedNotes[TPC_D_S];
+      computedNotes[TPC_G_B] = computedNotes[TPC_F_S];
+      computedNotes[TPC_A_B] = computedNotes[TPC_G_S];
+      computedNotes[TPC_B_B] = computedNotes[TPC_A_S];
+
+      fillMissing17();
+      }
+
+//---------------------------------------------------------
+//   fillMissing17
+//---------------------------------------------------------
+void Scale::fillMissing17() {
+      computedNotes[TPC_E_S] = computedNotes[TPC_F_B] = (computedNotes[TPC_E] + computedNotes[TPC_F]) / 2;
+      computedNotes[TPC_B_S] = computedNotes[TPC_C_B] = (computedNotes[TPC_B] + computedNotes[TPC_C]) / 2;
+
+      fillMissing21();
+      }
+
+//---------------------------------------------------------
+//   fillMissing21
+//---------------------------------------------------------
+void Scale::fillMissing21() {
+      computedNotes[TPC_B_SS] = computedNotes[TPC_D_BB] = computedNotes[TPC_C];
+      computedNotes[TPC_C_SS] = computedNotes[TPC_E_BB] = computedNotes[TPC_D];
+      computedNotes[TPC_D_SS] = computedNotes[TPC_F_BB] = computedNotes[TPC_E];
+      computedNotes[TPC_E_SS] = computedNotes[TPC_G_BB] = computedNotes[TPC_F];
+      computedNotes[TPC_F_SS] = computedNotes[TPC_A_BB] = computedNotes[TPC_G];
+      computedNotes[TPC_G_SS] = computedNotes[TPC_B_BB] = computedNotes[TPC_A];
+      computedNotes[TPC_A_SS] = computedNotes[TPC_C_BB] = computedNotes[TPC_B];
+      }
+
+//---------------------------------------------------------
+//   getTuning
+//---------------------------------------------------------
+
+double Scale::getTuning(const Note* note) const
+      {
+      static map<Tpc, double> standardNotes = {
+            {TPC_D_BB, 1200}, {TPC_C_S, 100},  {TPC_D_B, 100},  {TPC_C_SS, 200},  {TPC_D, 200},
+            {TPC_E_BB, 200},  {TPC_D_S, 300},  {TPC_E_B, 300},  {TPC_D_SS, 400},  {TPC_E, 400},
+            {TPC_F_BB, 400},  {TPC_E_S, 500},  {TPC_F_B, 400},  {TPC_E_SS, 500},  {TPC_F, 500},
+            {TPC_G_BB, 500},  {TPC_F_S, 600},  {TPC_G_B, 600},  {TPC_F_SS, 700},  {TPC_G, 700},
+            {TPC_A_BB, 700},  {TPC_G_S, 800},  {TPC_A_B, 800},  {TPC_G_SS, 900},  {TPC_A, 900},
+            {TPC_B_BB, 900},  {TPC_A_S, 1000}, {TPC_B_B, 1000}, {TPC_A_SS, 1100}, {TPC_B, 1100},
+            {TPC_C_BB, 1100}, {TPC_B_S, 1200}, {TPC_C_B, 1100}, {TPC_B_SS, 1200}, {TPC_C, 1200}};
+
+      auto standardNote = standardNotes.find((Tpc)note->tpc1());
+      auto computedNote = computedNotes.find((Tpc)note->tpc1());
+
+      if (standardNote == standardNotes.end() || computedNote == computedNotes.end()) {
+            return 0;
+            }
+
+      return standardNote->second - computedNote->second;
+      }
+}
```

Comment:
Your approach is more compact, but it is also harder to understand and thus maintain or modify.
Also what happens when we add new simbols that won't work with the logic, are we going to rewrite it completely?
Some issues I found : 
 - for TPC_FIRST_STEP = 3, TPC_STEPS_IN_FIFTH = 7, CENT_PER_OCTAVE = 1200
  tpc = TPC_F_BB = TPC_MIN + 1 the result would be :  (3 + 1 * 7) % 1200 = 10, but the result should be 400.
  - you're not support the 17 note variant
  - you should use 14 instead of TPC_DELTA_ENHARMONIC